### PR TITLE
Add: 2 Quirks Configs

### PR DESCRIPTION
**Pull request by Floofies**
## About The Pull Request

This PR adds two new config options for quirks:
- *Flag* `DISABLE_QUIRK_POINTS`
  - When enabled, disables quirk points balancing.
  - When enabled, players can select positive quirks without first selecting negative ones.
  - When enabled, the quirk points balance visually hides itself on the Quirks page.
- *Number* `MAX_POSITIVE_QUIRKS`
  - Limits the maximum quantity of positive quirks which players can select using the Character Preferences page.
  - I ported this from the old `MAX_QUIRKS` define.
  - When set to `0`, players won't be able to select any positive quirks, and they won't appear on the Quirks page.
  - When set to `-1`, players will be able to select any quantity of positive quirks.
  - When commented-out or undefined, the default is `6`.
  - When set to `0` or `-1`, the positive quirk balance visually hides itself on the Quirks page.

## Why It's Good For The Game

There is some downstream repositories asking for the quirks system to be configurable. Additionally, I always find myself tweaking these values on my own private servers and I thought it would be nice to share my edits. Usually I was simply commenting-out sections of this code in order to get the same result, so it helps to have an official way to disable quirk points.

## Changelog

:cl: A.C.M.O.
config: Added two new config flags for quirks, DISABLE_QUIRK_POINTS and MAX_POSITIVE_QUIRKS.
/:cl:

### CODE REVIEW

**Comment by Mothblocks** [comment]
We are currently in a feature freeze https://github.com/tgstation/tgstation/pull/80731

**Comment by Mothblocks** [comment]
Nvm this isn't player facing anything

File: code/controllers/subsystem/processing/quirks.dm [hunk 1]
```diff
@@ -198,15 +204,15 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
...
+			if ((max_positive_quirks >= 0) && (positive_quirks.len == max_positive_quirks))
 				continue
 
 			positive_quirks[quirk_name] = value
 
 		balance += value
 		new_quirks += quirk_name
 
-	if (balance > 0)
+	if (points_enabled && (balance > 0))
```
Comment (Mothblocks) [inline code/controllers/subsystem/processing/quirks.dm:215]: Don't bother with parentheses

File: code/controllers/subsystem/processing/quirks.dm [hunk 2]
```diff
@@ -198,15 +204,15 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
 
 		var/value = initial(quirk.value)
 		if (value > 0)
-			if (positive_quirks.len == MAX_QUIRKS)
+			if ((max_positive_quirks >= 0) && (positive_quirks.len == max_positive_quirks))
```
Comment (Mothblocks) [inline code/controllers/subsystem/processing/quirks.dm:207]: Don't bother with parentheses

File: code/controllers/configuration/entries/game_options.dm [hunk 3]
```diff
@@ -445,3 +445,12 @@
...
 	key_mode = KEY_MODE_TEXT
 	value_mode = VALUE_MODE_NUM
+
+// Configs for the Quirk system
+/// Disables Quirk point balancing for the server and clients.
+/datum/config_entry/flag/disable_quirk_points
+
+/// The maximum amount of positive quirks one character can have at roundstart.
+/datum/config_entry/number/max_positive_quirks
+	default = -1
```
Comment (Mothblocks) [inline code/controllers/configuration/entries/game_options.dm:455]: This means that all existing servers are going to have unlimited quirks. You need to default this to 6

**Comment by Floofies** [comment]
Yeah I accidentally used the "add" changelog tag when I meant to use "config". So this isn't a feature addition, it's just a config tweak to existing functionality.

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx [hunk 4]
```diff
@@ -445,11 +446,15 @@ export function QuirksPage(props) {
...
+                  <Box fontSize="1.3em">
+                    {pointsEnabled ? 'Quirk Balance' : 'Quirks Selected'}
+                  </Box>
                 </Stack.Item>
 
                 <Stack.Item>
-                  <StatDisplay>{balance}</StatDisplay>
+                  <StatDisplay>
+                    {pointsEnabled ? balance : selectedQuirks.length}
+                  </StatDisplay>
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:457]: Woudl rather you remove this entirely, the number of quirks you have is not useful with no limit

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx [hunk 5]
```diff
@@ -396,7 +397,7 @@ export function QuirksPage(props) {
 
                 <Stack.Item>
                   <StatDisplay>
-                    {positiveQuirks} / {maxPositiveQuirks}
+                    {positiveQuirks}{maxPositiveQuirks >= 0 ? ` / ${maxPositiveQuirks}` : ""}
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:400]: Don't show anything if there is no limit, just a plain number of count isn't useful

File: code/controllers/subsystem/processing/quirks.dm [hunk 6]
```diff
@@ -137,19 +138,22 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
...
-			possible_quirks -= quirk
-			continue
-		if(!quirk_points[quirk] > 0) //positive only
-			possible_quirks -= quirk
-			continue
-		good_count++
-		score += quirk_points[quirk]
-		quirks_to_add += quirk
+	if(max_positive_quirks < 0)
+		max_positive_quirks = 6
```
Comment (Mothblocks) [inline code/controllers/subsystem/processing/quirks.dm:142]: Nit: Do this closer to the definition of max_positive_quirks

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx (hunk 4 above)
Comment (Floofies) [reply tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:457]: I need to put something there to keep the webpage aligned... Removing it caused the whole list of selected quirks to move upwards. Any ideas?

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx (hunk 5 above)
Comment (Floofies) [reply tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:400]: Same issue as the above review. If I remove those elements then the quirks list moves upwards and out of alignment with the other half of the page.

File: code/controllers/configuration/entries/game_options.dm (hunk 3 above)
Comment (Floofies) [reply code/controllers/configuration/entries/game_options.dm]: Fixed by commit [74b561c](https://github.com/tgstation/tgstation/pull/81033/commits/74b561cdc7435eae66da2f589f41ef6c5c3fdcc5)

File: code/controllers/subsystem/processing/quirks.dm (hunk 2 above)
Comment (Floofies) [reply code/controllers/subsystem/processing/quirks.dm]: Fixed by commit [74b561c](https://github.com/tgstation/tgstation/pull/81033/commits/74b561cdc7435eae66da2f589f41ef6c5c3fdcc5)

File: code/controllers/subsystem/processing/quirks.dm (hunk 1 above)
Comment (Floofies) [reply code/controllers/subsystem/processing/quirks.dm]: Fixed by commit [74b561c](https://github.com/tgstation/tgstation/pull/81033/commits/74b561cdc7435eae66da2f589f41ef6c5c3fdcc5)

File: code/controllers/subsystem/processing/quirks.dm (hunk 6 above)
Comment (Floofies) [reply code/controllers/subsystem/processing/quirks.dm]: Fixed by commit [74b561c](https://github.com/tgstation/tgstation/pull/81033/commits/74b561cdc7435eae66da2f589f41ef6c5c3fdcc5)

File: code/controllers/subsystem/processing/quirks.dm (hunk 2 above)
Comment (Floofies) [reply code/controllers/subsystem/processing/quirks.dm]: Fixed by commit [74b561c](https://github.com/tgstation/tgstation/pull/81033/commits/74b561cdc7435eae66da2f589f41ef6c5c3fdcc5)

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx (hunk 4 above)
Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:458]: Put an empty div with some height

File: code/modules/client/preferences.dm [hunk 7]
```diff
@@ -436,6 +436,8 @@ GLOBAL_LIST_EMPTY(preferences_datums)
 			.++
 
 /datum/preferences/proc/validate_quirks()
+	if(CONFIG_GET(flag/disable_quirk_points))
+		return
```
Comment (ZephyrTFA) [inline code/modules/client/preferences.dm:440]: you still need to check for quirks which break selection rules

File: config/game_options.txt [hunk 8]
```diff
@@ -546,4 +546,19 @@ NEUTRAL_STATION_TRAITS 2.5 1
 NEGATIVE_STATION_TRAITS 0 8
 NEGATIVE_STATION_TRAITS 1 4
 NEGATIVE_STATION_TRAITS 2 2
-NEGATIVE_STATION_TRAITS 3 1
\ No newline at end of file
+NEGATIVE_STATION_TRAITS 3 1
+
+##------------------------------------------------
+
+# Configuration for character Quirks.
```
Comment (ZephyrTFA) [inline config/game_options.txt:553]: get rid of this header

File: code/modules/client/preferences.dm (hunk 7 above)
Comment (Floofies) [reply code/modules/client/preferences.dm:440]: This proc is not responsible for such functionality, and only checks to ensure that the quirk balance is not a positive number.

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx (hunk 4 above)
Comment (Floofies) [reply tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx:458]: Ok, I will attempt doing that!

**Comment by Jacquerel** [comment]
neat, will let us revert some code downstream

File: config/game_options.txt (hunk 8 above)
Comment (Floofies) [reply config/game_options.txt]: Fixed by commit [a92046a](https://github.com/tgstation/tgstation/pull/81033/commits/a92046aa3bae9c31087217f30072e9a9c2be0d19)

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx (hunk 5 above)
Comment (Floofies) [reply tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx]: Fixed by commit [9bcd434](https://github.com/tgstation/tgstation/pull/81033/commits/9bcd4341d097e89ea709bd8bb10fe94cad472c31)

File: tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx (hunk 4 above)
Comment (Floofies) [reply tgui/packages/tgui/interfaces/PreferencesMenu/QuirksPage.tsx]: Fixed by commit [9bcd434](https://github.com/tgstation/tgstation/pull/81033/commits/9bcd4341d097e89ea709bd8bb10fe94cad472c31)

**Comment by Floofies** [comment]
Things look very pleasantly aligned now, I used a couple of Box elements which set their own margins. I've also managed to hide positive quirks from TGUI if you happen to disable them in the config.

image

image

**Comment by Floofies** [comment]
Tested all the aspects of the PR, including spawning with the quirks, and everything seems to function as expected.

**Review by Ghommie — APPROVED** [review]

**Pull request merged**